service/src/java/org/apache/hive/service/servlet/OTELExporter.java

```diff
@@ -0,0 +1,225 @@
+        queryIdToSpanMap.remove(hQueryId);
+        queryIdToTasksMap.remove(hQueryId);
+
+        historicalQueryId.add(hQueryId);
+      }
+
+      //For queries that already ended either before OTEL service started or in between OTEL loops
+      if (historicalQueryId.add(hQuery.getQueryDisplay().getQueryId())) {
+        rootspan = tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId() + " - completed")
+                .startSpan();
+        Context parentContext = Context.current().with(rootspan);
+
+        Span initSpan = tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId() + " - completed").setParent(parentContext).startSpan()
+                .setAttribute("queryID", hQuery.getQueryDisplay().getQueryId())
+                .setAttribute("queryString", hQuery.getQueryDisplay().getQueryString())
+                .setAttribute("Begin Time", hQuery.getBeginTime());
+        if (hQuery.getQueryDisplay().getErrorMessage() != null) {
+          initSpan.setAttribute("Error Message", hQuery.getQueryDisplay().getErrorMessage());
+        }
+        initSpan.end();
+
+        for (QueryDisplay.TaskDisplay task : hQuery.getQueryDisplay().getTaskDisplays()) {
+          parentContext = Context.current().with(rootspan);
+          tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId()+ " - " + task.getTaskId() + " - completed")
+                  .setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                  .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan()
+                  .end(task.getEndTime(), TimeUnit.MILLISECONDS);
+        }
+        rootspan.setAllAttributes(addQueryAttributes(hQuery)).end();
+      }
+    }
+    historicalQueryId.retainAll(historicalQueryIDs);
+  }
+
+  private AttributesMap addQueryAttributes(QueryInfo query){
+    AttributesMap attributes = AttributesMap.create(Long.MAX_VALUE, Integer.MAX_VALUE);
+    attributes.put(AttributeKey.stringKey("queryId"), query.getQueryDisplay().getQueryId());
+    attributes.put(AttributeKey.stringKey("QueryString"), query.getQueryDisplay().getQueryString());
+    attributes.put(AttributeKey.longKey("QueryStartTime"), query.getQueryDisplay().getQueryStartTime());
+    attributes.put(AttributeKey.longKey("End Time"), query.getEndTime());
+    attributes.put(AttributeKey.stringKey("Operation Id"), query.getOperationId());
+    attributes.put(AttributeKey.stringKey("Operation Log Location"), query.getOperationLogLocation());
+    attributes.put(AttributeKey.stringKey("Error Message"), query.getQueryDisplay().getErrorMessage());
+    attributes.put(AttributeKey.stringKey("explainPlan"), query.getQueryDisplay().getExplainPlan());
+    attributes.put(AttributeKey.stringKey("fullLogLocation"), query.getQueryDisplay().getFullLogLocation());
+    attributes.put(AttributeKey.stringKey("taskDisplays"), Joiner.on("\t").join(query.getQueryDisplay().getTaskDisplays()));
```

Comment:
this ain't required IMO, we are having spans for each task now

Reply:
Yes, removed taskDisplays attribute.